proposals/gc/MVP.md

```diff
@@ -94,21 +108,47 @@ In addition to the [existing rules](https://github.com/WebAssembly/function-refe
 * every type is a subtype of `any`
   - `t <: any`
 
+* `dataref` is a subtype of `eqref`
+  - `data <: eq`
+  - TODO: provide a way to make data types non-eq, especially immutable ones?
+
 * `i31ref` is a subtype of `eqref`
   - `i31 <: eq`
 
-* Any concrete type is a subtype of `eq` if its not a function
-  - `(type $t) <: eq`
+* Any concrete type is a subtype of either `data` or `func`
+  - `(type $t) <: data`
      - if `$t = <structtype>` or `$t = <arraytype>`
-     - or `$t = type rt` and `rt <: eq` (imports)
-  - TODO: provide a way to make data types non-eq, especially immutable ones
+     - or `$t = type ht` and `rt <: data` (imports)
+  - `(type $t) <: func`
+     - if `$t = <functype>`
+     - or `$t = type ht` and `rt <: func` (imports)
+
+Note: This creates a hierarchy of *abstract* Wasm heap types that looks as follows.
+```
+      any
+     /   \
+   eq    func
+  /  \
+i31  data
+```
+All *concrete* heap types (of the form `(type $t)`) are situated below either `data` or `func`.
+
+In addition, the abstract heap type `extern` is also a subtype of `any`.
+Its interpretation is defined by the host environment.
+It may contain additional host-defined types that are neither of the above three leaf type categories.
+It may also overlap with some or all of these categories, as would be observable by applying a classification instruction like `ref.is_func` to a value of type `externref`.
+The possible outcomes of such an operation hence depend on the host environment.
+(For example, in a JavaScript embedding, `externref` could be inhabited by all JS values -- which is a natural choice, because JavaScript is untyped; but some of its values are JS-side representations of Wasm values per the JS API, and those can also be observed as `data` or `func` references. Another possible interpretation could be that `data` is disjoint from `extern`, which would be determined by the coercions allowed by the JS API at the JS/Wasm boudary.)
```

Comment:
This would cause problems for code migration. That is, I imagine we would like it to be easy for a program component written in JS to switch to GC-Wasm (possibly with JS-decorated `rtt`s for JS interop). But the suggestion here would make it so that the switch would cause previously accepted coercions (in other program components) to `externref` to become invalid. This means a converted module would have to create JS proxies (rather than `rtt` decorations) around all of its escaping references in order to ensure backwards compatibility.

Reply:
You had my attention at

> we would like it to be easy for a program component written in JS to switch to GC-Wasm

and I'd like to note that this isn't exactly what we are gravitating towards currently for other reasons, in part because of your earlier comments elsewhere. Rather seems like we are doing a slingshot around Jupiter while still hoping that there'll remain a faint signal of code migration in deep space, so we can point at it and say "look, it's still there". 🪐 🚀 💥

Reply:
@dcodeIO I have implemented systems with a typed language (similar to GC-Wasm) and an untyped OO language (similar to JS) that interop efficiently and in a manner that supports code migration. One reason I posed [this extension](https://github.com/WebAssembly/call-tags/blob/master/proposals/call-tags/Overview.md#extension) to the Call Tags proposal is that we found this mechanism sped up method calls in untyped code when invoked on objects originating from typed code. I have not, however, written up how to add such efficient interop to GC-Wasm and imports/exports because some of the techniques are not compatible with the current MVP (or Post-MVP). So I share the goal you have in mind, but I am trying to achieve it by first getting the right foundations in place for supporting efficient interop.

Reply:
I see, thanks for doing that! Whenever you have something of interest to me in this domain, feel free to ping me :)

Reply:
I am referring to the (existing) coercions to `externref` in the JS API.

Reply:
You'd have to be a little more concrete. At the boundary, externref does not imply any observable coercion, nor would its semantics change.

Reply:
Kotlin currently compiles to JS, representing Kotlin objects as JS objects. That means those Kotlin-as-JS objects can currently be given to wasm programs as `externref` values. Now if a Kotlin program is compiled to GC-wasm instead (maybe with some JS decoration in the `rtt` for JS interop), then your suggestion above would mean that Kotlin-as-wasm objects would not be passable to wasm programs as `externref` values. Thus previously working interactions would break by switching to wasm even if one does the job of decorating the rtts of their wasm references with JS info for JS interop.

Reply:
If a module explicitly asks to import an externref then switching between host and Wasm won't generally work. But I don't know what's surprising about that, you'd get what you asked for. Of course, there is no point in ever doing that, anyref is what you want on module boundaries.

Reply:
This is a surprising response. You're saying that the following is fine with you. Say you have a web program P using a third-party library L both currently implemented in JS. Suppose P changes to use wasm but in a backwards-compatible manner, using `externref` to handle JS values (such as L's) so that wasmP + L behaves as before. Suppose L changes to use wasm but in a backwards-compatible manner, using `externref` to handle JS values (such as P's) so that P + wasmL behaves as before. Then wasmP + wasmL (with JS glue) will not work because their uses of `externref` can no longer denote the other's values. That seems to make wasm interop with the web ecosystem rather fragile.

Reply:
I would assume both wasmL and wasmP would be using `externref` at the boundary, as they expect to be compatible with JavaScript on the other end. This would seem to "just work" if I understand correctly?

Reply:
Ah, I see now -- if I import a function from another module I have to specify the arguments' types on the import. If I specify `externref`, I can't call the other module's function with a GC value if it's now impossible to cast to `externref`.

A workaround could be to import a host function for each type that returns its argument as an `externref`. This might be awkward and slow compared to casts...

Reply:
> Suppose L changes to use wasm but in a backwards-compatible manner, using `externref` to handle JS values

L would not be able to provide Wasm data values as externref under that alternative semantics. This would not be a possible change, let alone a backwards-compatible one. If you truly said externref you should mean it. You are clearly leaking your implementation choice if you do.

Again, externref is not the type you ever want to use as an abstract type at the boundary -- if you leak implementation details, you are artificially restricting your evolution paths, and get what you ask for. You always want to use type imports or anyref, which properly abstract away a module's implementation choice.

Reply:
According to the current JS API for the MVP, a wasm module can export wasm references that have well-defined interactions in JS code. So wasmL can use wasm references where L used JS references, and if done right the change will be backwards compatible with other JS code, such as P. So wasmL imports `externref` values to reference (foreign) JS values, but exports reference values that the JS API "converts" (without wrapping) to JS values.

> Again, externref is not the type you ever want to use as an abstract type at the boundary

The reference types proposal says that "Any JS value can be passed as externref to a Wasm function, stored in a global, or in a table". So people are generating programs with the expectation that `externref` represents any JS value. Changing that to "Another possible interpretation could be that `data` is disjoint from `extern`" would make it difficult for new wasm programs to interact with old wasm programs in the web ecosystem that were not written with this distinction in mind.

Reply:
> You always want to use type imports or anyref, which properly abstract away a module's implementation choice.

* If I should always use `anyref`, why does `externref` exist?
* How do you reference JavaScript duck typing behavior in a type import?

Reply:
Also to clarify; if a JS value containing a Wasm GC struct or array reference is passed to a Wasm function that defines an argument as type `externref`, what happens? Does it trap now?

Reply:
Note that it makes a lot of sense to use `externref` for JS objects, since you might have to call methods and access properties on them and this is a JS-specific interface that requires importing helper functions like `Reflect.get`, `Reflect.construct`, and `Reflect.apply` from JavaScript, which will accept JavaScript values and so be imported using `externref` as their arguments and return values. There's little benefit I can imagine to choosing `anyref` over `externref` there.

If your API doesn't involve properties or prototype-bound methods on the JS objects, but only involves passing the objects as arguments to functions which you can import in Wasm, then maybe you wouldn't want to make it `externref`.

Or am I missing something and there's some other way to access regular JavaScript objects with properties and prototype-bound methods that doesn't involve calling out to JavaScript from Wasm?

Reply:
@brion:
> * If I should always use anyref, why does externref exist?
Externref is a wart that only came to be because we removed subtyping from the reference-types proposal at the last minute and thus anyref had to go.

> * How do you reference JavaScript duck typing behavior in a type import?
Sorry, I don't understand the question. There is no duck typing in Wasm.

> if a JS value containing a Wasm GC struct or array reference is passed to a Wasm function that defines an argument as type externref, what happens? Does it trap now?
The most likely interpretation would be that it is "coerced" in some form.

> Note that it makes a lot of sense to use externref for JS objects, since you might have to call methods and access properties on them and this is a JS-specific interface that requires importing helper functions like Reflect.get, Reflect.construct, and Reflect.apply from JavaScript, which will accept JavaScript values and so be imported using externref as their arguments and return values. There's little benefit I can imagine to choosing anyref over externref there.
All this works just fine with anyref. However, if you really want to explicitly use JS objects in such an interface, then you are indeed not interested in abstraction or the ability to transition them to Wasm objects instead. Then externref seems fine, and datarefs being disjoint would not be relevant.

@RossTate:
> According to the current JS API for the MVP
If you mean the GC MVP, then under a hypothetical design where dataref and externref were separate, it obviously would have to be adjusted accordingly.

Folks, before we waste more energy on this: this parenthesis was meant as an illustrative remark. I am not actually proposing to do it that way. It merely illustrates the design space that exists for hosts to define the meaning of externref. I realise that JS is not the greatest example to use, but it is the one that everybody knows. I added a clarifying sentence to point that out and (hopefully) avoid possible confusion.

Reply:
> All this works just fine with anyref. However, if you really want to explicitly use JS objects in such an interface, then you are indeed not interested in abstraction or the ability to transition them to Wasm objects instead. Then externref seems fine, and datarefs being disjoint would not be relevant.

The scenario is that we have two existing JavaScript modules which both provide JavaScritp APIs that interact with each other.

So yes, the APIs need to remain JavaScript-compatible when implementations are changed from JavaScript to Wasm.

That is why there are questions about how to implement that kind of code migration behavior, since "code migration" is a significant goal.

Reply:
I should clarify that the APIs already exist in this scenario! They were not designed with anything Wasm in mind; they are JS code that already exists.